test/utils/ExternalMulticallSimulator.sol.ts

```diff
@@ -0,0 +1,359 @@
+import type { HardhatEthersSigner } from '@nomicfoundation/hardhat-ethers/signers';
+import * as helpers from '@nomicfoundation/hardhat-network-helpers';
+import { expect } from 'chai';
+import type { AddressLike, BigNumberish, BytesLike, Interface, VoidSigner } from 'ethers';
+import { ethers } from 'hardhat';
+
+import { type DapiProxyWithOev, DapiProxyWithOev__factory } from '../../src/index';
+import * as testUtils from '../test-utils';
+
+interface Call {
+  target: AddressLike;
+  data: BytesLike;
+}
+
+async function getDapiTransmutationCalldata(
+  preTransmutationCalls: Call[],
+  postTransmutationCalls: Call[],
+  api3ServerV1Address: AddressLike,
+  api3ServerV1Interface: Interface,
+  dapiName: BytesLike,
+  value: BigNumberish,
+  externalMulticallSimulatorInterface: Interface
```

Comment:
I find this a bit too complex. I'd separate this into two parts:
1. Function to create the call data for dAPI transmutation
2. The function that simulates the multicall and decodes the response.

See: https://github.com/api3dao/contracts/commit/284a823696f6f10458dbcafed628b84de5647fe7

Reply:
I think making the provider call for the user is not great. I would have these functions instead
- `getDapiTransmutationCalls` (returns only two Calls for the transmutation)
- `encodeDapiTransmutationCalldata` (takes all Calls as argument, returns fully encoded calldata)
- The user is expected to use the calldata to make the call however they will
- `decodeReturnData` (takes returndata and all Calls as argument)

Call would include
- Target address
- [Human readable ABI](https://blog.ricmoo.com/human-readable-contract-abis-in-ethers-js-141902f4d917)
- Input args

which would allow us to decode the returndata

Reply:
That works as well. We can separate the encoding & sending transaction, but imo we should also provide the function that makes the staticcall (accepts RPC provider and the calldata).

I like the `which would allow us to decode the returndata`.